compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -1006,14 +1006,17 @@ object desugar {
    *           - None              :  sel @unchecked
    *           - Exhaustive        :  sel
    *           - IrrefutablePatDef,
-   *             IrrefutableGenFrom:  sel @unchecked with attachment `CheckIrrefutable -> checkMode`
+   *             IrrefutableGenFrom:  sel with attachment `CheckIrrefutable -> checkMode`
    */
   def makeSelector(sel: Tree, checkMode: MatchCheck)(using Context): Tree =
     if (checkMode == MatchCheck.Exhaustive) sel
     else {
-      val sel1 = Annotated(sel, New(ref(defn.UncheckedAnnot.typeRef)))
-      if (checkMode != MatchCheck.None) sel1.pushAttachment(CheckIrrefutable, checkMode)
-      sel1
+      if (checkMode == MatchCheck.None)
+        Annotated(sel, New(ref(defn.UncheckedAnnot.typeRef)))
+      else
+        // TODO: use `pushAttachment` and investigate duplicate attachment
+        sel.withAttachment(CheckIrrefutable, checkMode)
+        sel
```

Comment:
I cannot foresee the interaction of existing attachments here. However, the implementation of `makeSelector` could be factored to

```
checkMode match {
      case MatchCheck.Exhaustive => 
        sel
      case MatchCheck.None =>  
        Annotated(sel, New(ref(defn.UncheckedAnnot.typeRef)))      
      case MatchCheck.IrrefutablePatDef | MatchCheck.IrrefutableGenFrom => 
        // TODO: use `pushAttachment` and investigate duplicate attachment
        sel.withAttachment(CheckIrrefutable, checkMode)
    }
```
to be even closer to the comment documenting it.